src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+          Seq(if (TaskContext.get.partitionId == 0) Some(vw.getModel) else None).iterator
+        }.get // this will throw if there was an exception
+    }
+
+    val encoder = Encoders.kryo[Option[Array[Byte]]]
+
+    // schedule multiple mapPartitions in
+    val outerNumPasses = if (getEnableCacheFile) 1 else if (getNumPasses > 1 && !getCacheRows) getNumPasses else 1
+    var localInitialModel = if (isDefined(initialModel)) Some(getInitialModel) else None
+
+    for (p <- 0 until outerNumPasses) {
+      // dispatch to exectuors and collect the model of the first partition (everybody has the same at the end anyway)
+      localInitialModel = df.mapPartitions(inputRows => trainIteration(inputRows, localInitialModel, p))(encoder)
+        .reduce((a, b) => if (a.isEmpty) b else a)
+    }
+
+    localInitialModel
+  }
+
+  /**
+    * Setup spanning tree and invoke training.
+    * @param df input data.
+    * @param vwArgs VW command line arguments.
+    * @param numWorkers number of target workers.
+    * @return
+    */
+  protected def trainInternalDistributed(df: DataFrame, vwArgs: StringBuilder, numWorkers: Int) = {
+    // multiple partitions -> setup distributed coordination
+    val spanningTree = new ClusterSpanningTree(0, getArgs.contains("--quiet"))
+
+    try {
+      spanningTree.start
+
+      val jobUniqueId = Math.abs(UUID.randomUUID.getLeastSignificantBits.toInt)
+      val driverHostAddress = ClusterUtil.getDriverHost(df)
+      val port = spanningTree.getPort
+
+      /*
+      --span_server specifies the network address of a little server that sets up spanning trees over the nodes.
+      --unique_id should be a number that is the same for all nodes executing a particular job and
+        different for all others.
+      --total is the total number of nodes.
+      --node should be unique for each node and range from {0,total-1}.
+      --holdout_off should be included for distributed training
+      */
+      vwArgs.append(s" --holdout_off --span_server $driverHostAddress --span_server_port $port ")
+        .append(s"--unique_id $jobUniqueId --total $numWorkers ")
+
+      trainInternal(df, vwArgs.result, () => s"--node ${TaskContext.get.partitionId}")
+    } finally {
+      spanningTree.stop
+    }
+  }
+
+  /**
+    * Main training loop
+    * @param dataset input data.
+    * @return binary VW model.
+    */
+  protected def trainInternal(dataset: Dataset[_]): Array[Byte] = {
+    // follow LightGBM pattern
+    val numCoresPerExec = ClusterUtil.getNumCoresPerExecutor(dataset)
+    val numExecutorCores = ClusterUtil.getNumExecutorCores(dataset, numCoresPerExec)
+    val numWorkers = min(numExecutorCores, dataset.rdd.getNumPartitions)
+
+    // Reduce number of partitions to number of executor cores
+    val df = if (dataset.rdd.getNumPartitions > numWorkers)
+        dataset.toDF.coalesce(numWorkers)
+      else
+        dataset.toDF
+
+    // add exposed parameters to the final command line args
```

Comment:
are all of these params mandatory in VW?

Reply:
no, I'm only adding them if supplied...